app/concerns/reset_tenant_id.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+module ResetTenantId
+  extend ActiveSupport::Concern
+
+  class Configurator
+    include Singleton
+    attr_reader :definitions
+
+    def initialize
+      @definitions = {}
+    end
+
+    def define(klass,  &block)
+      definition = definitions[klass] ||= Definition.new(klass)
+      definition.instance_exec(&block)
+    end
+  end
+
+  class Definition
+    attr_reader :queries, :klass
+    delegate :each_value, to: :queries
+
+    def initialize(klass)
+      @klass = klass
+      @queries = {}
+    end
+
+    def updates(name, options = {}, &block)
+      @queries[name.to_sym] = Query.new(klass, name, options, &block)
+    end
+
+    def [](name)
+      @queries[name.to_sym]
+    end
+  end
+
+  class Query
+    module FindEach
+      def find_each(*options, &block)
+        defined?(super) ? super(*options, &block) : each(&block)
+      end
+    end
+
+    attr_reader :klass, :name, :update, :collection
+
+    def initialize(klass, name, options = {}, &block)
+      @klass = klass
+      @name = name.to_sym
+      @collection = wrap_collection(options[:collection] || @name)
+      @update = wrap_update(options[:with] || block)
+    end
+
+    private
+
+    # :nodoc:
+    # Returns an object responding to #call(record)
+    def wrap_update(with)
+      case with
+      when NilClass
+        raise ArgumentError, 'Pass in `:with` option or a block'
+      when Symbol
+        proc do |record|
+          record.update_column :tenant_id, record.public_send(with) # rubocop:disable Rails/SkipsModelValidations
```

Comment:
I think the API is confusing. When you pass a `Symbol`, it will call that symbol (`public_send`) on the `record`, so the value that the symbol "returns"(*) is the value moved to `tenant_id`. When you pass something that responds to `call`, then it's _not_ the value returned by the proc that will be used to set `tenant_id` but it becomes responsibility of the proc to set the `tenant_id`. In other words, `with` can be an alternative to passing a block (a complex operation that is expected to update the record), but it can also be the name of a method that will just return a scalar value.

I would expect `with` always to resolve to a scalar value. It can be a symbol or a proc, but in both cases the value returning value will be used with `record.update_column :tenant_id, ...`. And if you want to override `record.update_column :tenant_id, ...` then you have to pass a block and forget about `with`.

(*) Actually it's the value that _the method named equal to the symbol_ returns.
______

In a nutshell, something like:

```ruby
def wrap_update(with)
  raise ArgumentError, 'Pass in `:with` option or a block' if with.blank?

  proc do |record|
    record.update_column :tenant_id, with.is_a?(Symbol) ? record.public_send(with) : with.call
  end
end
```

Reply:
Well it is the same as in rails :) I do not see that confusing.

`Symbol#to_proc`

Reply:
It was intended by default it would use update_column but if you want to do something more particular then it will use the block or the with.